pkg/ds/farm.go

```diff
@@ -0,0 +1,411 @@
+func (dsf *Farm) dsContends(dsFields *ds_fields.DaemonSet) (ds_fields.ID, bool, error) {
+	// This daemon set does not contend if it is disabled
+	if dsFields.Disabled {
+		return "", false, nil
+	}
+	// Get all eligible nodes for this daemon set by looking at the labes.NODE tree
+	eligibleNodes, err := dsf.scheduler.EligibleNodes(dsFields.Manifest, dsFields.NodeSelector)
+	if err != nil {
+		return "", false, util.Errorf("Error retrieving eligible nodes for daemon set: %v", err)
+	}
+
+	// If this daemon set has a node selector set to Everything, check the labels
+	// of other daemon sets
+	for _, child := range dsf.children {
+		everythingSelector := klabels.Everything().String()
+		if !child.ds.IsDisabled() && child.ds.PodID() == dsFields.PodID && child.ds.ID() != dsFields.ID {
+			// Naively check if both selectors are the Everything selector plus
+			// something else or if both selectors are the same
+			//
+			// This will still think that the following two selectors contend:
+			// { az = zone_one, az != zone_one } and {} (the everything selector)
+			// even though the first selector doesn't select anything
+			//
+			// If either the child or the current daemon set has the everything selector
+			// then they contend
+			if dsFields.NodeSelector.String() == everythingSelector ||
+				child.ds.GetNodeSelector().String() == everythingSelector {
+				dsf.raiseContentionAlert(child.ds.ID(), dsFields.ID)
+				return child.ds.ID(), true, nil
+			}
+
+			// If both daemon sets have the same selector, then they contend
+			//
+			// This will still think that the following two selectors contend:
+			// { az = zone_one, az != zone_one } and { az = zone_one, az != zone_one }
+			// even though they don't select anything
+			if dsFields.NodeSelector.String() == child.ds.GetNodeSelector().String() {
+				dsf.raiseContentionAlert(child.ds.ID(), dsFields.ID)
+				return child.ds.ID(), true, nil
+			}
+
+			// Check the child's eligibleNodes, then intersect it to see if there
+			// are any overlaps
+			//
+			// NOTE: This is naive, it does not account for new nodes, so any alerts
+			// we get will be caused by adding new nodes by both human, machine error,
+			// are starting up a daemon set farm where contention already exists
+			scheduledNodes, err := child.ds.EligibleNodes()
+			if err != nil {
+				return "", false, util.Errorf("Error getting scheduled nodes: %v", err)
+			}
+			intersectedNodes := types.NewNodeSet(eligibleNodes...).Intersection(types.NewNodeSet(scheduledNodes...))
+			if intersectedNodes.Len() > 0 {
+				dsf.raiseContentionAlert(child.ds.ID(), dsFields.ID)
+				return child.ds.ID(), true, nil
+			}
+		}
+	}
+
+	return "", false, nil
+}
+
+func (dsf *Farm) raiseContentionAlert(oldDSID ds_fields.ID, newDSID ds_fields.ID) {
+	if alertErr := dsf.alerter.Alert(alerting.AlertInfo{
+		Description: fmt.Sprintf("New ds '%v', contends with '%v'", newDSID, oldDSID),
+		IncidentKey: "preemptive_ds_contention",
```

Comment:
consider a `Details` struct with pod id, availability zone, and other information for diagnostics